plugin/iamutil/dataset_resource.go

```diff
@@ -0,0 +1,145 @@
+package iamutil
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/go-gcp-common/gcputil"
+)
+
+// NOTE: BigQuery does not conform to the typical REST for IAM policies
+// instead it has an access array with bindings on the dataset
+// object. https://cloud.google.com/bigquery/docs/reference/rest/v2/datasets#Dataset
+type AccessBinding struct {
+	Role         string `json:"role,omitempty"`
+	UserByEmail  string `json:"userByEmail,omitempty"`
+	GroupByEmail string `json:"groupByEmail,omitempty"`
+}
+
+type Dataset struct {
+	Access []*AccessBinding `json:"access,omitempty"`
+	Etag   string           `json:"etag,omitempty"`
+}
+
+// NOTE: DatasetResource implements IamResource.
+// This is because bigquery datasets have their own
+// ACLs instead of an IAM policy
+type DatasetResource struct {
+	relativeId *gcputil.RelativeResourceName
+	config     *RestResource
+}
+
+func (r *DatasetResource) GetConfig() *RestResource {
+	return r.config
+}
+
+func (r *DatasetResource) GetRelativeId() *gcputil.RelativeResourceName {
+	return r.relativeId
+}
+
+func (r *DatasetResource) GetIamPolicy(ctx context.Context, h *ApiHandle) (*Policy, error) {
+	var dataset Dataset
+	if err := h.DoGetRequest(ctx, r, &dataset); err != nil {
+		return nil, errwrap.Wrapf("unable to get BigQuery Dataset ACL: {{err}}", err)
+	}
+	p := datasetAsPolicy(&dataset)
+	return p, nil
+}
+
+func (r *DatasetResource) SetIamPolicy(ctx context.Context, h *ApiHandle, p *Policy) (*Policy, error) {
+	var jsonP []byte
+	ds, err := policyAsDataset(p)
+	if err != nil {
+		return nil, err
+	}
+	jsonP, err = json.Marshal(ds)
+	if err != nil {
+		return nil, err
+	}
+	reqJson := fmt.Sprintf(r.config.SetMethod.RequestFormat, jsonP)
+	if !json.Valid([]byte(reqJson)) {
+		return nil, fmt.Errorf("request format from generated BigQuery Dataset config invalid JSON: %s", reqJson)
+	}
+
+	var dataset Dataset
+	if err := h.DoSetRequest(ctx, r, strings.NewReader(reqJson), &dataset); err != nil {
+		return nil, errwrap.Wrapf("unable to set BigQuery Dataset ACL: {{err}}", err)
+	}
+	policy := datasetAsPolicy(&dataset)
+
+	return policy, nil
+}
+
+func policyAsDataset(p *Policy) (*Dataset, error) {
+	ds := &Dataset{
+		Etag: p.Etag,
+	}
+
+	if p == nil {
```

Comment:
This check should be moved to the top, since currently line 79 will panic.

Reply:
Good catch!